platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+     * for adoption.
+     * A Node SHALL NOT, ever, have more than one candidate roster
+     * at the same time.
+     */
+    byte candidate_roster_hash = 1;
+
+    /**
+     * A list of round numbers and roster hashes.
+     * The round number indicates the round in which the corresponding roster became active
+     * <p>
+     * This list SHALL be ordered by round numbers.<br/>
+     * This list MUST be deterministically ordered.
+     */
+    repeated RoundRosterPair round_roster_pairs = 2;
+}
+```
+
+#### RoundRosterPair
+
+A `RoundRosterPair` will be defined as follows:
+
+```proto
+/**
+ * A pair of round number and active roster hash.
+ * <p>
+ * This message SHALL encapsulate the round number and the hash of the
+ * active roster used for that round.
+ */
+message RoundRosterPair {
+
+    /**
+     * The round number.
+     * <p>
+     * This value SHALL be the round number of the consensus round in which this roster became active.
+     */
+    uint64 round_number = 1;
+
+    /**
+     * The SHA-384 hash of the active roster for the given round number.
+     * <p>
+     * This value SHALL be the hash of the active roster used for the round.
+     */
+    byte active_roster_hash = 2;
+}
+```
+
+It is noteworthy that the roster must be immutable to guarantee the integrity of the computed hash.
+This map of rosters will typically contain the current Active Roster, an optional previously Active Roster, and an
+optional current Candidate Roster. Insertion of rosters will be controlled by ensuring that the acceptance of a new
+Candidate Roster invalidates and removes the current candidate roster. Therefore, the map is expected to have a maximum
+size of three elements.
+The new `RosterState` will store the `candidateRosterHash` and a list `roundRosterPairs`
+of pairs of round numbers and the active roster that was used for that round. The `roundRosterPairs` list will be
+updated only when the active roster changes.
+In the vast majority of cases, the list will contain only one element, the current active roster.
+Immediately following the adoption of a candidate roster, this list will have two elements — the previous active roster
+and the current active roster (which was the candidate roster).
+The choice of a list is suitable here as it is much cheaper to iterate over 2-3 elements comparing their `long` round
+numbers than it is to do a hash lookup.
+When we reach full Dynamic Address Book, this design choice may need to be revisited.
+
+If a `candidateRosterHash` hash entry already exist (hash collisions) in the map of Rosters, it will be discarded. That
+is, setting a candidate roster is an idempotent operation.
+
+One benefit of this indirection (using a Map instead of a Singleton) is that it avoids moving data around in the merkle
+state.
+Another benefit is that adoption trigger becomes straightforward (App sets the roster) with delineated
+responsibilities between the App and the platform.
```

Comment:
I don't understand this design. We have a virtual map "Roster Map" with hashOfRoster->Roster. And we have RosterState (which is a singleton?) with the hash of the current roster and then a list of round/hashOfRoster as to what round the roster became active in.

First, what is the advantage of the RosterMap? Why not just have RosterState contain the actual candidate roster and list of round/roster? Why have it be hashes?

Second, if there are only ever going to be 3 items in the RosterMap, why use a map and not just a singleton with 3 fields?

Third, the management of the list of round_roster_pairs sounds complex. What are the theoretical minimum and maximum number of entries in this list?

Reply:
Answering the first question last:

2. `RosterMap` is the durable storage for all rosters that must be stored in the state, regardless of their type. It is expected to be a few items, yes, but not necessarily 3. In theory, in the full DAB case, we could have multiple previous active rosters between all the current non-ancient rounds (4? 5? 6?). Using a singleton with n fields means the entire thing has to be re-hashed each time (polluting the record stream?). So we opted for a state data structure that avoids that, a map.

3. Before DAB, the round_roster_pairs list will have two items. The item at index 0 will be the active roster hash and the round it became active, index 1 will be for the previous active roster. Most times, the list will be of size 1. The period immediately following an upgrade, it will have 2 items. After DAB, the theoretical size is much higher than 2, but realistically less than say, 10. The consideration here is that since we know index 0 will always be the active roster, it’s cheaper to just get element at that known index. We could replace all of these with singletons, but it takes us farther away from DAB implementation, not closer.

Now, back to question 1
1. The existence of the round number in the `RosterState` makes it slightly faster changing than the RosterMap, which means it will need more rehashing. If this is correct, and we move the Roster into it, we will have to re-hash the rosters many more times than the current design requires. The Roster is a fairly ‘big’ data structure. Rehashing it many times over is not desirable. That’s why we opted to separate them.
If we get rid of the RosterMap and instead store all Rosters in RosterState, it will come with a higher re-hash tax.

I’ve set up a call to discuss these in details tomorrow with @Jasperpotts.